### PR TITLE
Add TFCLIPModel

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Add `TFCLIPModel`, along with `TFCLIPTextModel` and `TFCLIPVisionModel`.

(and fixed doc examples in `modeling_clip.py`)

## Who can review?
@Rocketknight1 @patrickvonplaten @NielsRogge @sgugger @LysandreJik

### CODE REVIEW

**Comment by ydshieh** [comment]
Finally finalized this PR after working on something else. It is ready for review :)

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 1]
```diff
@@ -0,0 +1,1528 @@
...
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        training: bool = False,
+        **kwargs,
+    ) -> Union[TFBaseModelOutputWithPooling, Tuple[tf.Tensor]]:
+        inputs = input_processing(
+            func=self.call,
+            config=self.config,
+            input_ids=pixel_values,
```
Comment (patil-suraj) [inline src/transformers/models/clip/modeling_tf_clip.py:724]: I'm not familiar with this. But this seems nit hackish and as there are going to be more vision models IMO it should be handled explicitly rather than setting `input_ids=pixel_values`. cc @Rocketknight1

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 2]
```diff
@@ -0,0 +1,1528 @@
...
+            return_dict=return_dict,
+            training=training,
+            kwargs_call=kwargs,
+        )
+
+        if "input_ids" in inputs:
+            inputs["pixel_values"] = inputs.pop("input_ids")
+
+        if inputs["pixel_values"] is None:
+            raise ValueError("You have to specify pixel_values")
```
Comment (patil-suraj) [inline src/transformers/models/clip/modeling_tf_clip.py:736]: We should probably check this before `input_processing`

File: tests/test_modeling_tf_vit.py [hunk 3]
```diff
@@ -200,7 +200,7 @@ def test_model(self):
 
     # overwrite from common since `encoder_seq_length` and `encoder_key_length` are calculated
     # in a different way than in text models.
-    @tooslow
+    @slow
```
Comment (patil-suraj) [inline tests/test_modeling_tf_vit.py:203]: IMO this should be changed in a separate PR.

File: src/transformers/models/clip/modeling_clip.py [hunk 4]
```diff
@@ -827,7 +827,7 @@ def forward(
 
             >>> outputs = model(**inputs)
             >>> last_hidden_state = outputs.last_hidden_state
-            >>> pooled_output = outputs.pooled_output # pooled CLS states
+            >>> pooled_output = outputs.pooler_output # pooled CLS states
```
Comment (patil-suraj) [inline src/transformers/models/clip/modeling_clip.py:830]: Good catch!

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for adding the TF version @ydshieh!

All the CLIP-related logic looks good to me! @Rocketknight1  Would be awesome if you could review the TF code :)

File: src/transformers/activations_tf.py [hunk 5]
```diff
@@ -63,6 +63,12 @@ def gelu_fast(x):
     return 0.5 * x * (1.0 + tf.tanh(x * coeff2 * (1.0 + coeff1 * x * x)))
 
 
+def quick_gelu(x):
```
Comment (Rocketknight1) [inline src/transformers/activations_tf.py:66]: I really want to delete all of these approximate functions and just not support TF <= 2.3! (But let's leave that for another PR)

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 6]
```diff
@@ -0,0 +1,1528 @@
...
+        logits_per_text:(:obj:`tf.Tensor` of shape :obj:`(text_batch_size, image_batch_size)`):
+            The scaled dot product scores between :obj:`text_embeds` and :obj:`image_embeds`. This represents the
+            text-image similarity scores.
+        text_embeds(:obj:`tf.Tensor` of shape :obj:`(batch_size, output_dim`):
+            The text embeddings obtained by applying the projection layer to the pooled output of
+            :class:`~transformers.TFCLIPTextModel`.
+        image_embeds(:obj:`tf.Tensor` of shape :obj:`(batch_size, output_dim`):
+            The image embeddings obtained by applying the projection layer to the pooled output of
+            :class:`~transformers.TFCLIPVisionModel`.
+        text_model_output(:obj:`TFBaseModelOutputWithPooling`):
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:109]:
```suggestion
        text_model_output(:class:`~transformers.modeling_tf_utils.TFBaseModelOutputWithPooling`):
```

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 7]
```diff
@@ -0,0 +1,1528 @@
...
+            The output of the :class:`~transformers.TFCLIPTextModel`.
+        vision_model_output(:obj:`TFBaseModelOutputWithPooling`):
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:111]:
```suggestion
        vision_model_output(:class:`~transformers.modeling_tf_utils.TFBaseModelOutputWithPooling`):
```

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 8]
```diff
@@ -0,0 +1,1528 @@
...
+        super().build(input_shape)
+
+    def call(self, pixel_values: tf.Tensor) -> tf.Tensor:
+
+        batch_size, num_channels, height, width = shape_list(pixel_values)
+
+        # When running on CPU, `tf.nn.conv2d` doesn't support `NCHW` format.
+        # So change the input format from `NCHW` to `NHWC`.
+        # shape = (batch_size, in_height, in_width, in_channels=num_channels)
+        pixel_values = tf.transpose(pixel_values, perm=(0, 2, 3, 1))
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:181]: Should we properly document we expect inputs in the NCHW format? I think TF users might be more used to NHWC?

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 9]
```diff
@@ -0,0 +1,1528 @@
...
+            data_format="channels_last",
+            use_bias=False,
+            kernel_initializer=get_initializer(self.config.initializer_range),
+            name="patch_embedding",
+        )
+
+    def build(self, input_shape: tf.TensorShape):
+
+        # Q: should we use normal initializer?
+        # initializer = tf.random_normal_initializer(mean=0.0, stddev=1.0)
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:157]: Needs to be addressed @patil-suraj do you have an answer?

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 10]
```diff
@@ -0,0 +1,1528 @@
...
+        attention_output = tf.matmul(attention_probs, value_layer)
+        attention_output = tf.transpose(attention_output, perm=[0, 2, 1, 3])
+
+        # (batch_size, seq_len_q, embed_dim)
+        attention_output = tf.reshape(tensor=attention_output, shape=(batch_size, -1, self.embed_dim))
+
+        attention_output = self.out_proj(attention_output, training=training)
+        # In TFBert, attention weights are returned after dropout.
+        # However, in CLIP, they are returned before dropout.
+        # Q: What to do?
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:346]: This should be consistent with the PyTorch version.

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 11]
```diff
@@ -0,0 +1,1528 @@
...
+        return_dict: bool,
+        training: bool = False,
+        **kwargs,
+    ) -> Union[TFBaseModelOutputWithPooling, Tuple[tf.Tensor]]:
+        input_shape = shape_list(input_ids)
+
+        embedding_output = self.embeddings(
+            input_ids=input_ids,
+            position_ids=position_ids,
+        )
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:505]:
```suggestion
        embedding_output = self.embeddings(input_ids=input_ids, position_ids=position_ids)
```
Fits in one line.

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 12]
```diff
@@ -0,0 +1,1528 @@
...
+@keras_serializable
+class TFCLIPTextMainLayer(tf.keras.layers.Layer):
+    config_class = CLIPTextConfig
+
+    def __init__(self, config: CLIPTextConfig, **kwargs):
+        super().__init__(**kwargs)
+
+        self.config = config
+
+        self.text_model = TFCLIPTextTransformer(config, name="text_model")
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:594]: No need for all those new lines IMO.

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 13]
```diff
@@ -0,0 +1,1528 @@
...
+@keras_serializable
+class TFCLIPVisionMainLayer(tf.keras.layers.Layer):
+    config_class = CLIPVisionConfig
+
+    def __init__(self, config: CLIPVisionConfig, **kwargs):
+        super().__init__(**kwargs)
+
+        self.config = config
+
+        self.vision_model = TFCLIPVisionTransformer(config, name="vision_model")
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:707]: Same here: no need for all the new lines.

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 14]
```diff
@@ -0,0 +1,1528 @@
...
+
+        Returns:
+            :obj:`Dict[str, tf.Tensor]`: The dummy inputs.
+        """
+        VISION_DUMMY_INPUTS = tf.random.uniform(
+            shape=(len(DUMMY_INPUTS), 3, self.config.image_size, self.config.image_size), dtype=tf.float32
+        )
+        return {
+            "pixel_values": VISION_DUMMY_INPUTS,
+        }
```
Comment (sgugger) [inline src/transformers/models/clip/modeling_tf_clip.py:1220]:
```suggestion
        return {"pixel_values": VISION_DUMMY_INPUTS}
```

File: tests/test_modeling_tf_common.py [hunk 15]
```diff
@@ -276,12 +276,19 @@ def test_onnx_runtime_optimize(self):
...
+            inputs_dict.pop("return_loss", None)
+
         tf_main_layer_classes = set(
             module_member
             for model_class in self.all_model_classes
             for module in (import_module(model_class.__module__),)
             for module_member_name in dir(module)
             if module_member_name.endswith("MainLayer")
+            # This condition is required, since `modeling_tf_clip.py` has 3 classes whose names end with `MainLayer`.
+            and module_member_name[: -len("MainLayer")] == model_class.__name__[: -len("Model")]
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:291]: Can we group the two together with a single
```
if module_member_name.replace("MainLayer", "Model") == model_class.__name__
```

File: tests/test_modeling_tf_common.py [hunk 16]
```diff
@@ -276,12 +276,19 @@ def test_onnx_runtime_optimize(self):
     def test_keras_save_load(self):
         config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
 
+        # `CLIPModelTester/TFCLIPModelTester` set `return_loss` to `True`, however this causes the preparation
+        # of `symbolic_inputs` below failed.
+        if self.__class__.__name__ == "TFCLIPModelTest":
+            inputs_dict.pop("return_loss", None)
+
```
Comment (sgugger) [inline tests/test_modeling_tf_common.py:283]: This should not be in the common tests, but the clip tester should overwrite this test.

**Review by sgugger — APPROVED** [review]
Thanks a lot for adding this model!

File: src/transformers/models/clip/modeling_tf_clip.py (hunk 8 above)
Comment (Rocketknight1) [reply src/transformers/models/clip/modeling_tf_clip.py:181]: @ydshieh I discussed this with Sylvain, and we decided keeping NCHW is definitely better!

File: src/transformers/models/clip/modeling_tf_clip.py (hunk 9 above)
Comment (patil-suraj) [reply src/transformers/models/clip/modeling_tf_clip.py:157]: Aah yes! CLIP's initialization logic is bit complex, indeed we should use `normal` here.
See the ` CLIPPreTrainedModel.__init_weights` method for reference.

https://github.com/huggingface/transformers/blob/4cdb67caba9abee8b5b94b4f88f55e95a3d1013f/src/transformers/models/clip/modeling_clip.py#L345

File: src/transformers/models/clip/modeling_clip.py (hunk 4 above)
Comment (patil-suraj) [reply src/transformers/models/clip/modeling_clip.py:830]: This is fixed in #14641

File: src/transformers/models/clip/modeling_tf_clip.py (hunk 10 above)
Comment (Rocketknight1) [reply src/transformers/models/clip/modeling_tf_clip.py:346]: +1 on this - match the PyTorch CLIP implementation!

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 17]
```diff
@@ -0,0 +1,1528 @@
...
+            logits_per_image=logits_per_image,
+            logits_per_text=logits_per_text,
+            text_embeds=text_embeds,
+            image_embeds=image_embeds,
+            text_model_output=text_outputs,
+            vision_model_output=vision_outputs,
+        )
+
+
+class TFCLIPPreTrainedModel(TFPreTrainedModel):
```
Comment (Rocketknight1) [inline src/transformers/models/clip/modeling_tf_clip.py:977]: The PyTorch version of this class also included [the weight initializations for all of the layers in the model](https://github.com/huggingface/transformers/blob/master/src/transformers/models/clip/modeling_clip.py#L345-L386) - I agree that this probably isn't the right place to put them in TF, but we should make sure that we use the same initialization values.

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 18]
```diff
@@ -0,0 +1,1528 @@
...
+        if inputs_embeds is None:
+            inputs_embeds = tf.gather(params=self.weight, indices=input_ids)
+
+        input_shape = shape_list(inputs_embeds)[:-1]
+
+        if position_ids is None:
+            position_ids = tf.expand_dims(tf.range(start=0, limit=input_shape[-1]), axis=0)
+
+        position_embeds = tf.gather(params=self.position_embedding, indices=position_ids)
+        position_embeds = tf.tile(input=position_embeds, multiples=(input_shape[0], 1, 1))
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/models/clip/modeling_tf_clip.py:251]: @Rocketknight1 Good point. Yes, it is possible. However

- The line `position_embeds = tf.tile(input=position_embeds, multiples=(input_shape[0], 1, 1))` is everywhere in TF model scripts. If we change here, probably it is better to change all (?)
- `tf.broadcast_to` is not perfect as mentioned in TF doc [broadcast_to](https://www.tensorflow.org/api_docs/python/tf/broadcast_to)

  ```
  When doing broadcasted operations such as multiplying a tensor by a scalar,
  broadcasting (usually) confers some time or space benefit, as the broadcasted
  tensor is never materialized.
  
  However, broadcast_to does not carry with it any such benefits. The newly-created
  tensor takes the full memory of the broadcasted shape. (In a graph context, broadcast_to
  might be fused to subsequent operation and then be optimized away, however.)
  ```

But after diving deeper, I found something more severe: 

The `...INPUTS_DOCSTRING` (everywhere in transformers) specifies `position_ids` should have shape `(batch_size, sequence_length)`. If a user specify `position_ids` manually with the full `batch_size` (very unlikely though), then the line `tf.tile` shouldn't be run! (otherwise the `batch_size` will become squared). 

If we use `tf.broadcast_to`, we don't have such issue.

Could you discuss with @sgugger  and @LysandreJik  to see what to do about this, please?

**Review by Rocketknight1 — APPROVED** [review]
Overall, this is a great PR! I have a couple of nits in addition to the comments raised by other people, but the main one is that I think layer initializations don't actually match the ones from the PyTorch model, because I don't think the values from the `_init_weights` function in `CLIPPreTrainedModel` are present anywhere in the TF file. Obviously, this isn't a problem for pretrained models, but it's definitely worth fixing for anyone who wants to train their own model from scratch.

**Review by ydshieh — COMMENTED** [review]
I just saw that all my previous own reviews are in pending status.
(I was thinking why I didn't have any reply ...)

I have answered some questions and raised some questions too.
I tag some of you in each review comments, hope not too spamming.

File: tests/test_modeling_tf_common.py (hunk 15 above)
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:291]: if `module_member_name` itself is something like `TFCLIP(Text/Vision)Model`, the condition in your comment will hold, but we only want to deal with `TFCLIP(Text/Vision)MainLayer`, right?

File: tests/test_modeling_tf_common.py (hunk 16 above)
Comment (ydshieh) [reply tests/test_modeling_tf_common.py:283]: OK

File: src/transformers/models/clip/modeling_tf_clip.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/clip/modeling_tf_clip.py:251]: I don't think we need `tf.broadcast_to` either - because the first dimension of `position_embeds` is 1, we can just remove the tile line entirely, and position_embeds will still be correctly broadcast in the next line. In other words, the code would just be:
```
position_embeds = tf.gather(params=self.position_embedding, indices=position_ids)
final_embeddings = inputs_embeds + position_embeds
```

I did some quick testing and this seems to work fine. However, it's just a nit, and the performance impact will likely be small anyway, so don't worry about it too much! We can also just leave it, and I'll make a note later to check the codebase for instances of `tile` that should be replaced with implicit broadcasting.

[the remark this answers is not in the document]
Comment (Rocketknight1) [reply src/transformers/models/clip/modeling_tf_clip.py:251]: Also, thinking about it more, using broadcasting resolves your issue with the users specifying a full batch of `position_ids`, so maybe it's more than just a minor performance nit!

[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/models/clip/modeling_tf_clip.py:251]: > 

Thanks, @Rocketknight1 

Well, implicit broadcasting will still work when users specifying a full batch - there won't be any broadcasting in this case.

But as you said, we can leave it to another PR if necessary :)

File: src/transformers/models/clip/modeling_tf_clip.py (hunk 8 above)
Comment (ydshieh) [reply src/transformers/models/clip/modeling_tf_clip.py:181]: `CLIP_VISION_INPUTS_DOCSTRING` has already mentioned

```
pixel_values: ... and each example must have the shape :obj:`(batch_size, num_channels, height, width)`):
```
I just added (for `TFCLIPVisionEmbeddings`)
```
  def call(self, pixel_values: tf.Tensor) -> tf.Tensor:
      """`pixel_values` is expected to be of NCHW format."""
```

File: src/transformers/models/clip/modeling_tf_clip.py [hunk 19]
```diff
@@ -0,0 +1,1528 @@
...
+
+        batch_size, seq_length = input_shape
+        # CLIP's text model uses causal mask, prepare it here.
+        # https://github.com/openai/CLIP/blob/cfcffb90e69f37bf2ff1e988237a0fbe41f33c04/clip/model.py#L324
+        causal_attention_mask = self._build_causal_attention_mask(batch_size, seq_length, dtype=embedding_output.dtype)
+
+        # # expand attention_mask
+        # if attention_mask is not None:
+        #     # [bsz, seq_len] -> [bsz, 1, tgt_seq_len, src_seq_len]
+        #     attention_mask = _expand_mask(attention_mask, hidden_states.dtype)
```
[the remark this answers is not in the document]
Comment (ydshieh) [reply src/transformers/models/clip/modeling_tf_clip.py]: I was mixing/using another way to expand the mask. I changed it back to use `_expand_mask` which aligns with `CLIPTextTransformer`

**Comment by ydshieh** [comment]
I addressed all the reviews, in particular the weight initialization.

In PT version, there is 

```
if isinstance(module, CLIPTextEmbeddings):
    module.token_embedding.weight.data.normal_(mean=0.0, std=factor * 0.02)
    module.position_embedding.weight.data.normal_(mean=0.0, std=factor * 0.02)
```

I am not sure why `0.02` is used instead of `config.initializer_range`. cc @patil-suraj

**Comment by LysandreJik** [comment]
Should this be merged @Rocketknight1 @patil-suraj? :smiley:

**Comment by ydshieh** [comment]
> Should this be merged @Rocketknight1 @patil-suraj? 😃

There is (new) conflicts I just saw. I will resolve later today or tomorrow.

**Comment by patil-suraj** [comment]
okay for merge on my side (once the conflicts are resolved) !

**Comment by sgugger** [comment]
Merging will require:
- a rebase on master
- converting all docstrings to Markdown.

You can install our `doc-builder` utility to do this:
```
pip install git+https://github.com/huggingface/doc-builder
```

Then run from your branch in the Transformers repo
```
doc-builder convert src/transformers/models/clip/modeling_tf_clip.py
```
to do this in the newly added modeling_tf_clip file.

Note that clip.rst has disappeared and you will need to add your model to clip.mdx instead.

Let us know if you need any help and please ping for review when you're done!

**Comment by ydshieh** [comment]
@patil-suraj @Rocketknight1 @sgugger I rebased on master and fixed errors. The last failed test is irrelevant I think.
Ready to be merged when you have some time :)

Thank you for reviewing this PR!

**Comment by Rocketknight1** [comment]
@sgugger I'm happy at this point!

**Review by sgugger — APPROVED** [review]
Thanks a lot for making all those changes!

Failures are unrelated, so merging.
